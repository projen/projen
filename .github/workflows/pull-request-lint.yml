# ~~ Generated by projen. To modify, edit .projenrc.ts and run "node ./projen.js".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
  merge_group: {}
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target')
    steps:
      - uses: amannn/action-semantic-pull-request@v6
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
  contributorStatement:
    name: Require Contributor Statement
    runs-on: ubuntu-latest
    permissions:
      pull-requests: read
    env:
      PR_BODY: ${{ github.event.pull_request.body }}
      EXPECTED: By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
      HELP: Contributor statement missing from PR description. Please include the following text in the PR description
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target') && !(github.event.pull_request.user.login == 'cdklabs-automation' || github.event.pull_request.user.login == 'dependabot[bot]')
    steps:
      - uses: actions/github-script@v8
        with:
          script: |-
            const actual = process.env.PR_BODY.replace(/\r?\n/g, "\n");
            const expected = process.env.EXPECTED.replace(/\r?\n/g, "\n");
            if (!actual.includes(expected)) {
                console.log("%j", actual);
                console.log("%j", expected);
                core.setFailed(`${process.env.HELP}: ${expected}`);
            }
