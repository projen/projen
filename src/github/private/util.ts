export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

export function context(value: string) {
  return `\${{ ${value} }}`;
}

// Checks if part of the file path is hidden
export function isHiddenPath(path: string) {
  return /(^|\/)\.[^\/\.]/g.test(path);
}

// Helper to assert a path is not hidden
export function ensureNotHiddenPath(value: string, name: string) {
  if (isHiddenPath(value)) {
    throw Error(`${name} cannot be a hidden path, got: ${value}`);
  }
}
