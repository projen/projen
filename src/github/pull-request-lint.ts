import { GitHub, PullRequestTemplate } from ".";
import { Job, JobPermission } from "./workflows-model";
import { Component } from "../component";
import { GroupRunnerOptions, filteredRunsOnOptions } from "../runner-options";

/**
 * Options for PullRequestLint
 */
export interface PullRequestLintOptions {
  /**
   * Validate that pull request titles follow Conventional Commits.
   *
   * @default true
   * @see https://www.conventionalcommits.org/
   */
  readonly semanticTitle?: boolean;

  /**
   * Options for validating the conventional commit title linter.
   * @default - title must start with "feat", "fix", or "chore"
   */
  readonly semanticTitleOptions?: SemanticTitleOptions;

  /**
   * Github Runner selection labels
   * @default ["ubuntu-latest"]
   * @description Defines a target Runner by labels
   * @throws {Error} if both `runsOn` and `runsOnGroup` are specified
   */
  readonly runsOn?: string[];

  /**
   * Github Runner Group selection options
   * @description Defines a target Runner Group by name and/or labels
   * @throws {Error} if both `runsOn` and `runsOnGroup` are specified
   */
  readonly runsOnGroup?: GroupRunnerOptions;

  /**
   * Require a contributor statement to be included in the PR description.
   * For example confirming that the contribution has been made by the contributor and complies with the project's license.
   *
   * Appends the statement to the end of the Pull Request template.
   *
   * @default - no contributor statement is required
   */
  readonly contributorStatement?: string;

  /**
   * Options for requiring a contributor statement on Pull Requests
   * @default - none
   */
  readonly contributorStatementOptions?: ContributorStatementOptions;
}

/**
 * Options for linting that PR titles follow Conventional Commits.
 * @see https://www.conventionalcommits.org/
 */
export interface SemanticTitleOptions {
  /**
   * Configure a list of commit types that are allowed.
   * @default ["feat", "fix", "chore"]
   */
  readonly types?: string[];

  /**
   * Configure that a scope must always be provided.
   * e.g. feat(ui), fix(core)
   * @default false
   */
  readonly requireScope?: boolean;

  /**
   * Configure which scopes are allowed (newline-delimited).
   * These are regex patterns auto-wrapped in `^ $`.
   *
   * @default - all scopes allowed
   */
  readonly scopes?: string[];
}

/**
 * Options for requiring a contributor statement on Pull Requests
 */
export interface ContributorStatementOptions {
  /**
   * Pull requests from these GitHub users are exempted from a contributor statement.
   * @default - no users are exempted
   */
  readonly exemptUsers?: string[];
  /**
   * Pull requests with one of these labels are exempted from a contributor statement.
   * @default - no labels are excluded
   */
  readonly exemptLabels?: string[];
}

/**
 * Configure validations to run on GitHub pull requests.
 * Only generates a file if at least one linter is configured.
 */
export class PullRequestLint extends Component {
  constructor(
    private readonly github: GitHub,
    private readonly options: PullRequestLintOptions = {}
  ) {
    super(github.project);

    const checkSemanticTitle = options.semanticTitle ?? true;
    const checkContributorStatement = Boolean(options.contributorStatement);

    // should only create a workflow if one or more linters are enabled
    if (!checkSemanticTitle && !checkContributorStatement) {
      return;
    }

    const workflow = github.addWorkflow("pull-request-lint");
    workflow.on({
      pullRequestTarget: {
        types: [
          "labeled",
          "opened",
          "synchronize",
          "reopened",
          "ready_for_review",
          "edited",
        ],
      },
      // run on merge group, but use a condition later to always succeed
      // needed so the workflow can be a required status check
      mergeGroup: {},
    });

    // All checks are run against the PR and can only be evaluated within a PR context
    // Needed so jobs can be set as required and will run successfully on merge group checks.
    const prCheck =
      "(github.event_name == 'pull_request' || github.event_name == 'pull_request_target')";

    if (checkSemanticTitle) {
      const opts = options.semanticTitleOptions ?? {};
      const types = opts.types ?? ["feat", "fix", "chore"];

      const validateJob: Job = {
        name: "Validate PR title",
        if: prCheck,
        ...filteredRunsOnOptions(options.runsOn, options.runsOnGroup),
        permissions: {
          pullRequests: JobPermission.WRITE,
        },
        steps: [
          {
            uses: "amannn/action-semantic-pull-request@v5.4.0",
            env: {
              GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}",
            },
            with: {
              types: types.join("\n"),
              ...(opts.scopes ? { scopes: opts.scopes.join("\n") } : {}),
              requireScope: opts.requireScope ?? false,
            },
          },
        ],
      };

      workflow.addJobs({ validate: validateJob });
    }

    if (options.contributorStatement) {
      const opts = options.contributorStatementOptions ?? {};
      const users = opts.exemptUsers ?? [];
      const labels = opts.exemptLabels ?? [];

      const conditions = [prCheck];

      const exclusions: string[] = [
        ...labels.map(
          (l) => `contains(github.event.pull_request.labels.*.name, '${l}')`
        ),
        ...users.map((u) => `github.event.pull_request.user.login == '${u}'`),
      ];

      if (exclusions.length) {
        conditions.push(`!(${exclusions.join(" || ")})`);
      }

      const script = (core: any) => {
        const actual = process.env.PR_BODY!.replace(/\r?\n/g, "\n");
        const expected = process.env.EXPECTED!.replace(/\r?\n/g, "\n");
        if (!actual.includes(expected)) {
          console.log("%j", actual);
          console.log("%j", expected);
          core.setFailed(`${process.env.HELP}: ${expected}`);
        }
      };

      const helpMessage =
        "Contributor statement missing from PR description. Please include the following text in the PR description";
      const contributorStatement: Job = {
        name: "Require Contributor Statement",
        runsOn: options.runsOn ?? ["ubuntu-latest"],
        permissions: {
          pullRequests: JobPermission.READ,
        },
        if: conditions.join(" && "),
        env: {
          PR_BODY: "${{ github.event.pull_request.body }}",
          EXPECTED: options.contributorStatement,
          HELP: helpMessage,
        },
        steps: [
          {
            uses: "actions/github-script@v6",
            with: {
              script: fnBody(script),
            },
          },
        ],
      };

      workflow.addJobs({ contributorStatement });
    }
  }

  public preSynthesize(): void {
    if (this.options.contributorStatement) {
      // Append to PR template in preSynthesize so it's always at the end of the file
      const prTemplate =
        PullRequestTemplate.of(this.project) ??
        this.github.addPullRequestTemplate();
      prTemplate?.addLine("");
      prTemplate?.addLine("---");
      prTemplate?.addLine(this.options.contributorStatement);
      prTemplate?.addLine("");
    }
  }
}

/**
 * Helper to generate a JS script as string from a function object
 * @returns A prettified string of the function's body
 */
function fnBody(fn: (...args: any[]) => any) {
  const def = fn.toString().replace(/\r?\n/g, "\n");
  const body = def
    .substring(def.indexOf("{") + 1, def.lastIndexOf("}"))
    .split("\n");
  const minIndentation = Math.min(
    ...body
      .filter((l) => l.trim()) // ignore empty lines
      .map((l) => l.search(/\S|$/))
  );

  return body
    .map((l) => l.replace(" ".repeat(minIndentation), ""))
    .join("\n")
    .trim();
}
