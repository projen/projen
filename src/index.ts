export * from "./ai-instructions";
export * from "./compare";
export * from "./component";
export * from "./dev-env";
export * from "./dependencies";
export * from "./docker-compose";
export * from "./dockerfile";
export * from "./file";
export * from "./gitattributes";
export * from "./gitpod";
export * from "./runner-options";
export * from "./ignore-file";
export * from "./ini";
export * from "./json";
export * from "./json-patch";
export * from "./logger";
export * from "./license";
export * from "./makefile";
export * from "./object-file";
export * from "./option-hints";
export * from "./project";
export * from "./project-build";
export * from "./project-tree";
export * from "./projects";
export * from "./projenrc";
export * from "./projenrc-json";
export * from "./readme";
export * from "./renovatebot";
export * from "./sample-file";
export * from "./semver";
export * from "./source-code";
export * from "./task";
export * from "./tasks";
export * from "./task-model";
export * from "./task-runtime";
export * from "./testing";
export * from "./textfile";
export * from "./toml";
export * from "./version";
export * from "./yaml";
export * from "./xmlfile";

// export submodules
export * as web from "./web";
export * as github from "./github";
export * as gitlab from "./gitlab";
export * as circleci from "./circleci";
export * as vscode from "./vscode";
export * as java from "./java";
export * as python from "./python";
export * as typescript from "./typescript";
export * as javascript from "./javascript";
export * as release from "./release";
export * as awscdk from "./awscdk";
export * as cdk8s from "./cdk8s";
export * as cdktf from "./cdktf";
export * as cdk from "./cdk";
export * as build from "./build";
