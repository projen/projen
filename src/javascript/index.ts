export * from "./bundler";
export * from "./eslint";
export * from "./jest";
export * from "./license-checker";
export * from "./node-package";
export * from "./node-project";
export * from "./npm-config";
export * from "./prettier";
export * from "./projenrc";
export * from "./typescript-config";
export * from "./upgrade-dependencies";
export * from "./yarnrc";
