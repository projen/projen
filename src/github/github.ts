import { GitHubActionsProvider } from "./actions-provider";
import { Dependabot, DependabotOptions } from "./dependabot";
import { GithubCredentials } from "./github-credentials";
import { MergeQueue, MergeQueueOptions } from "./merge-queue";
import { Mergify, MergifyOptions } from "./mergify";
import { PullRequestTemplate } from "./pr-template";
import {
  PullRequestBackport,
  PullRequestBackportOptions,
} from "./pull-request-backport";
import { PullRequestLint, PullRequestLintOptions } from "./pull-request-lint";
import { GithubWorkflow } from "./workflows";
import { Component } from "../component";
import { Project } from "../project";

export interface GitHubOptions {
  /**
   * Whether mergify should be enabled on this repository or not.
   *
   * @default true
   */
  readonly mergify?: boolean;

  /**
   * Options for Mergify.
   *
   * @default - default options
   */
  readonly mergifyOptions?: MergifyOptions;

  /**
   * Whether a merge queue should be used on this repository to merge pull requests.
   * Requires additional configuration of the repositories branch protection rules.
   *
   * @default false
   */
  readonly mergeQueue?: boolean;

  /**
   * Options for MergeQueue.
   *
   * @default - default options
   */
  readonly mergeQueueOptions?: MergeQueueOptions;

  /**
   * Enables GitHub workflows. If this is set to `false`, workflows will not be created.
   *
   * @default true
   */
  readonly workflows?: boolean;

  /**
   * Add a workflow that allows backport of PRs to other branches using labels.
   * When opening a new PR add a backport label to it,
   * and the PR will be backported to the target branches once the PR is merged.
   *
   * Should not be used together with mergify.
   *
   * @default false
   */
  readonly pullRequestBackport?: boolean;

  /**
   * Options for configuring pull request backport.
   *
   * @default - see defaults in `PullRequestBackportOptions`
   */
  readonly pullRequestBackportOptions?: PullRequestBackportOptions;

  /**
   * Add a workflow that performs basic checks for pull requests, like
   * validating that PRs follow Conventional Commits.
   *
   * @default true
   */
  readonly pullRequestLint?: boolean;

  /**
   * Options for configuring a pull request linter.
   *
   * @default - see defaults in `PullRequestLintOptions`
   */
  readonly pullRequestLintOptions?: PullRequestLintOptions;

  /**
   * Choose a method of providing GitHub API access for projen workflows.
   *
   * @default - use a personal access token named PROJEN_GITHUB_TOKEN
   */
  readonly projenCredentials?: GithubCredentials;

  /**
   * The name of a secret which includes a GitHub Personal Access Token to be
   * used by projen workflows. This token needs to have the `repo`, `workflows`
   * and `packages` scope.
   *
   * @default "PROJEN_GITHUB_TOKEN"
   * @deprecated - use `projenCredentials`
   */
  readonly projenTokenSecret?: string;

  /**
   * Download files in LFS in workflows
   *
   * @default true if the associated project has `lfsPatterns`, `false` otherwise
   */
  readonly downloadLfs?: boolean;
}

export class GitHub extends Component {
  /**
   * Returns the `GitHub` component of a project or `undefined` if the project
   * does not have a GitHub component.
   */
  public static of(project: Project): GitHub | undefined {
    const isGitHub = (c: Component): c is GitHub => c instanceof GitHub;
    return project.components.find(isGitHub);
  }

  /**
   * The `Mergify` component configured on this repository
   * This is `undefined` if Mergify is not enabled for this repository.
   */
  public readonly mergify?: Mergify;

  /**
   * The `MergeQueue` component configured on this repository
   * This is `undefined` if merge queues are not enabled for this repository.
   */
  public readonly mergeQueue?: MergeQueue;

  /**
   * Are workflows enabled?
   */
  public readonly workflowsEnabled: boolean;

  /**
   * GitHub API authentication method used by projen workflows.
   */
  public readonly projenCredentials: GithubCredentials;

  /**
   * The GitHub Actions provider used to manage the versions of actions used in steps
   */
  public readonly actions: GitHubActionsProvider;

  private readonly _downloadLfs?: boolean;

  public constructor(project: Project, options: GitHubOptions = {}) {
    super(project);

    this.actions = new GitHubActionsProvider();

    this.workflowsEnabled = options.workflows ?? true;

    this._downloadLfs = options.downloadLfs;

    if (options.projenCredentials && options.projenTokenSecret) {
      throw new Error(
        "projenTokenSecret is deprecated, please use projenCredentials instead"
      );
    }

    if (options.projenTokenSecret) {
      this.projenCredentials = GithubCredentials.fromPersonalAccessToken({
        secret: options.projenTokenSecret,
      });
    } else if (options.projenCredentials) {
      this.projenCredentials = options.projenCredentials;
    } else {
      this.projenCredentials = GithubCredentials.fromPersonalAccessToken({
        secret: "PROJEN_GITHUB_TOKEN",
      });
    }

    if (options.mergify ?? true) {
      this.mergify = new Mergify(this, options.mergifyOptions);
    }

    if (options.mergeQueue ?? false) {
      this.mergeQueue = new MergeQueue(this, options.mergeQueueOptions);
    }

    if (options.pullRequestLint ?? true) {
      new PullRequestLint(this, options.pullRequestLintOptions);
    }

    if (options.pullRequestBackport ?? false) {
      if (options.mergify) {
        this.project.logger.warn(
          "pullRequestBackport should not be used with mergify as mergify provides its own backport functionality. Please disable pullRequestBackport or mergify."
        );
      }
      new PullRequestBackport(this, options.pullRequestBackportOptions);
    }
  }

  /**
   * All workflows.
   */
  public get workflows(): GithubWorkflow[] {
    const isWorkflow = (c: Component): c is GithubWorkflow =>
      c instanceof GithubWorkflow;
    return this.project.components
      .filter(isWorkflow)
      .sort((w1, w2) => w1.name.localeCompare(w2.name));
  }

  /**
   * Adds a workflow to the project.
   * @param name Name of the workflow
   * @returns a GithubWorkflow instance
   */
  public addWorkflow(name: string) {
    const workflow = new GithubWorkflow(this, name);
    return workflow;
  }

  public addPullRequestTemplate(...content: string[]) {
    return new PullRequestTemplate(this, { lines: content });
  }

  public addDependabot(options?: DependabotOptions) {
    return new Dependabot(this, options);
  }

  /**
   * Finds a GitHub workflow by name. Returns `undefined` if the workflow cannot be found.
   * @param name The name of the GitHub workflow
   */
  public tryFindWorkflow(name: string): undefined | GithubWorkflow {
    return this.workflows.find((w) => w.name === name);
  }

  /**
   * Whether downloading from LFS is enabled for this GitHub project
   */
  public get downloadLfs() {
    return this._downloadLfs ?? this.project.gitattributes.hasLfsPatterns;
  }
}
