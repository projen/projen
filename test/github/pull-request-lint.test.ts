import * as yaml from "yaml";
import { PullRequestLint } from "../../src/github/pull-request-lint";
import { NodeProject, NodeProjectOptions } from "../../src/javascript";
import { synthSnapshot } from "../util";

test("default", () => {
  // GIVEN
  const project = createProject();

  // WHEN
  new PullRequestLint(project.github!);

  // THEN
  const snapshot = synthSnapshot(project);
  expect(snapshot[".github/workflows/pull-request-lint.yml"]).toBeDefined();

  const workflow = snapshot[".github/workflows/pull-request-lint.yml"];
  expect(workflow).toMatchSnapshot();
  expect(workflow).toContain("Validate PR title");
  expect(workflow).not.toContain("Require Contributor Statement");
});

describe("semantic titles", () => {
  test("configure types", () => {
    // GIVEN
    const project = createProject();

    // WHEN
    new PullRequestLint(project.github!, {
      semanticTitle: true,
      semanticTitleOptions: {
        types: ["feat", "fix"],
      },
    });

    // THEN
    const snapshot = synthSnapshot(project);
    expect(
      snapshot[".github/workflows/pull-request-lint.yml"]
    ).toMatchSnapshot();
  });

  test("require scope", () => {
    // GIVEN
    const project = createProject();

    // WHEN
    new PullRequestLint(project.github!, {
      semanticTitle: true,
      semanticTitleOptions: {
        requireScope: true,
      },
    });

    // THEN
    const snapshot = synthSnapshot(project);
    expect(
      snapshot[".github/workflows/pull-request-lint.yml"]
    ).toMatchSnapshot();
  });

  test("configure scopes", () => {
    // GIVEN
    const project = createProject();

    // WHEN
    new PullRequestLint(project.github!, {
      semanticTitle: true,
      semanticTitleOptions: {
        scopes: ["cli", "api"],
      },
    });

    // THEN
    const snapshot = synthSnapshot(project);
    expect(
      snapshot[".github/workflows/pull-request-lint.yml"]
    ).toMatchSnapshot();
  });
});

describe("contributor statement", () => {
  test("validates pull requests", () => {
    // GIVEN
    const project = createProject();

    // WHEN
    new PullRequestLint(project.github!, {
      contributorStatement: "Lorem ipsum dolor sit amet",
    });

    // THEN
    const snapshot = synthSnapshot(project);
    expect(
      snapshot[".github/workflows/pull-request-lint.yml"]
    ).toMatchSnapshot();
  });

  test("creates pull request template", () => {
    // GIVEN
    const project = createProject({
      pullRequestTemplate: false,
    });
    const contributorStatement = "Lorem ipsum dolor sit amet";

    // WHEN
    new PullRequestLint(project.github!, {
      contributorStatement,
    });

    // THEN
    const snapshot = synthSnapshot(project);
    expect(snapshot[".github/pull_request_template.md"]).toMatchSnapshot();
    expect(snapshot[".github/pull_request_template.md"]).toContain(
      contributorStatement
    );
  });

  test("appends to an existing request template", () => {
    // GIVEN
    const project = createProject({
      pullRequestTemplate: true,
      pullRequestTemplateContents: ["Foobar #"],
    });
    const contributorStatement = "Lorem ipsum dolor sit amet";

    // WHEN
    new PullRequestLint(project.github!, {
      contributorStatement,
    });

    // THEN
    const snapshot = synthSnapshot(project);
    expect(snapshot[".github/pull_request_template.md"]).toMatchSnapshot();
    expect(snapshot[".github/pull_request_template.md"]).toContain("Foobar #");
    expect(snapshot[".github/pull_request_template.md"]).toContain(
      contributorStatement
    );
  });

  test("can exempt users and labels", () => {
    // GIVEN
    const project = createProject();
    const contributorStatement = "Lorem ipsum dolor sit amet";

    // WHEN
    new PullRequestLint(project.github!, {
      contributorStatement,
      contributorStatementOptions: {
        exemptLabels: ["automation"],
        exemptUsers: ["github-bot[bot]"],
      },
    });

    // THEN
    const snapshot = synthSnapshot(project);
    expect(
      snapshot[".github/workflows/pull-request-lint.yml"]
    ).toMatchSnapshot();
    expect(snapshot[".github/workflows/pull-request-lint.yml"]).toContain(
      "github.event.pull_request.user.login == 'github-bot[bot]'"
    );
    expect(snapshot[".github/workflows/pull-request-lint.yml"]).toContain(
      "contains(github.event.pull_request.labels.*.name, 'automation')"
    );
  });
});

test("with custom runner", () => {
  // GIVEN
  const project = createProject();

  // WHEN
  new PullRequestLint(project.github!, {
    runsOn: ["self-hosted"],
  });

  // THEN
  const snapshot = synthSnapshot(project);
  expect(snapshot[".github/workflows/pull-request-lint.yml"]).toContain(
    "runs-on: self-hosted"
  );
});

test("with custom runner group", () => {
  // GIVEN
  const project = createProject();

  // WHEN
  new PullRequestLint(project.github!, {
    runsOnGroup: {
      group: "Default",
      labels: ["self-hosted", "x64", "linux"],
    },
  });

  // THEN
  const snapshot = synthSnapshot(project);
  const build = yaml.parse(snapshot[".github/workflows/pull-request-lint.yml"]);

  expect(build).toHaveProperty("jobs.validate.runs-on.group", "Default");
  expect(build).toHaveProperty("jobs.validate.runs-on.labels", [
    "self-hosted",
    "x64",
    "linux",
  ]);
});

type ProjectOptions = Omit<
  NodeProjectOptions,
  "outdir" | "defaultReleaseBranch" | "name"
>;
function createProject(options: ProjectOptions = {}): NodeProject {
  return new NodeProject({
    defaultReleaseBranch: "main",
    name: "node-project",
    githubOptions: {
      pullRequestLintOptions: {
        semanticTitle: false,
      },
    },
    ...options,
  });
}
