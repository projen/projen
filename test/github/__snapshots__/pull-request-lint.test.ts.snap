// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contributor statement appends to an existing request template 1`] = `
"Foobar #

---
Lorem ipsum dolor sit amet
"
`;

exports[`contributor statement can exempt users and labels 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
  merge_group: {}
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target')
    steps:
      - uses: amannn/action-semantic-pull-request@v5.4.0
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
  contributorStatement:
    name: Require Contributor Statement
    runs-on: ubuntu-latest
    permissions:
      pull-requests: read
    env:
      PR_BODY: \${{ github.event.pull_request.body }}
      EXPECTED: Lorem ipsum dolor sit amet
      HELP: Contributor statement missing from PR description. Please include the following text in the PR description
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target') && !(contains(github.event.pull_request.labels.*.name, 'automation') || github.event.pull_request.user.login == 'github-bot[bot]')
    steps:
      - uses: actions/github-script@v6
        with:
          script: |-
            const actual = process.env.PR_BODY.replace(/\\r?\\n/g, "\\n");
            const expected = process.env.EXPECTED.replace(/\\r?\\n/g, "\\n");
            if (!actual.includes(expected)) {
                console.log("%j", actual);
                console.log("%j", expected);
                core.setFailed(\`\${process.env.HELP}: \${expected}\`);
            }
"
`;

exports[`contributor statement creates pull request template 1`] = `
"Fixes #

---
Lorem ipsum dolor sit amet
"
`;

exports[`contributor statement validates pull requests 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
  merge_group: {}
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target')
    steps:
      - uses: amannn/action-semantic-pull-request@v5.4.0
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
  contributorStatement:
    name: Require Contributor Statement
    runs-on: ubuntu-latest
    permissions:
      pull-requests: read
    env:
      PR_BODY: \${{ github.event.pull_request.body }}
      EXPECTED: Lorem ipsum dolor sit amet
      HELP: Contributor statement missing from PR description. Please include the following text in the PR description
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target')
    steps:
      - uses: actions/github-script@v6
        with:
          script: |-
            const actual = process.env.PR_BODY.replace(/\\r?\\n/g, "\\n");
            const expected = process.env.EXPECTED.replace(/\\r?\\n/g, "\\n");
            if (!actual.includes(expected)) {
                console.log("%j", actual);
                console.log("%j", expected);
                core.setFailed(\`\${process.env.HELP}: \${expected}\`);
            }
"
`;

exports[`default 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
  merge_group: {}
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target')
    steps:
      - uses: amannn/action-semantic-pull-request@v5.4.0
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
"
`;

exports[`semantic titles configure scopes 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
  merge_group: {}
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target')
    steps:
      - uses: amannn/action-semantic-pull-request@v5.4.0
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          scopes: |-
            cli
            api
          requireScope: false
"
`;

exports[`semantic titles configure types 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
  merge_group: {}
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target')
    steps:
      - uses: amannn/action-semantic-pull-request@v5.4.0
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
          requireScope: false
"
`;

exports[`semantic titles require scope 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
  merge_group: {}
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target')
    steps:
      - uses: amannn/action-semantic-pull-request@v5.4.0
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: true
"
`;
